usr.sbin/slstats/slstats.8
.\"	@(#) $Header: /scm/cvs/src/usr.sbin/slstats/slstats.8,v 1.1 1995/10/18 08:48:21 deraadt Exp $ (LBL)
.Dd July 5, 1993
.Dt SLSTATS 8
.Sh NAME
.Nm slstats
.Nd report slip statistics
.Sh SYNOPSIS
.Nm slstats
.Op Fl v
.Op Fl i Ar interval
.Op Ar unit-number
.Oo
.Ar system
.Op Ar core
.Oc
.Sh DESCRIPTION
.Pp
The 
.Nm slstats
utility reports slip-related statistics for the
.Nm sl Ns
.So
.Ar unit-number
.Sc
interface.  If the
.Nm unit-number
is unspecified, it defaults to 0.  These statistics are displayed at
regular intervals.
.Pp
The display is split horizontally into input and output sections
containing columns of statistics describing the properties and volume
of packets received and transmitted by the specified interface.
.Pp
The first report will consist of a snapshot of the interface's present
statistics.  All further output will describe activity between report
intervals.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl v
Display additional statistics demonstrating the efficacy of VJ header
compression and providing more explicit information on failure
distribution.
.It Fl i
Specifies the interval between reports. The default interval is 5 seconds.
.It Ar system
Extract the name list from the specified system instead of the default
.Pa /netbsd .
.It Ar core
Extract values associated with the name list from the specified
.Ar core
instead of the default
.Pa /dev/kmem .
.El
.Pp
The following fields are printed on the input side:
.Bl -tag -width search
.It Li IN
The total number of bytes received by this interface.
.It Li PACK
The total number of packets received by this interface.
.It Li COMP
The number of compressed TCP packets received by this interface.
.It Li UNCOMP
The number of uncompressed TCP packets received by this interface.
.It Li ERR
The number of packets received by this interface of unknown type.
.It Li TOSS
The number of packets dropped on reception by this interface.  Only
reported when the
.Fl v
option is specified.
.It Li IP
The total number of non-TCP packets received by this interface. Only
reported when the
.Fl v
option is specified.
.El
.Pp
The following fields are printed on the output side:
.Bl -tag -width search
.It Li OUT
The total number of bytes transmitted from this interface.
.It Li PACK
The total number of packets transmitted from this interface.
.It Li COMP
The number of TCP packets compressed and then transmitted from this interface.
.It Li UNCOMP
The number of uncompressed TCP packets transmitted from this interface.
.It Li IP
The total number of non-TCP packets transmitted from this interface.
.It Li SEARCH
The number of searches for the cached header entry for a compressed
packet.  Only reported when the
.Fl v
option is specified.
.It Li MISS
The number of failed searches for the cached header entry for a
compressed packet.  Only reported when the
.Fl v
option is specified.
.El
.Sh BUGS
Should use
.Fl M Ar core
and
.Fl N Ar system
like most other programs using the kvm library.
.Pp
Should take an interface name and not a unit number.
.Pp
.Sh SEE ALSO
.Xr sliplogin 1


usr.sbin/slstats/slstats.c
/*
 * print serial line IP statistics:
 *	slstats [-i interval] [-v] [interface] [system] [core]
 *
 * Copyright (c) 1989, 1990, 1991, 1992 Regents of the University of
 * California. All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 *	Van Jacobson (van@ee.lbl.gov), Dec 31, 1989:
 *	- Initial distribution.
 */

#ifndef lint
/*static char rcsid[] =
    "@(#) $Header: /scm/cvs/src/usr.sbin/slstats/slstats.c,v 1.1 1995/10/18 08:48:21 deraadt Exp $ (LBL)";*/
static char rcsid[] = "$Id: slstats.c,v 1.1 1995/10/18 08:48:21 deraadt Exp $";
#endif

#include <stdio.h>
#include <paths.h>
#include <nlist.h>

#define INET

#include <sys/param.h>
#include <sys/mbuf.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/file.h>
#include <errno.h>
#include <signal.h>
#include <net/if.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_var.h>

#include <net/slcompress.h>
#include <net/if_slvar.h>

struct nlist nl[] = {
#define N_SOFTC 0
	{ "_sl_softc" },
	"",
};

char	*system = _PATH_UNIX;
char	*kmemf = _PATH_KMEM;

int	kflag;
int	vflag;
unsigned interval = 5;
int	unit;

extern	char *malloc();
extern	off_t lseek();

main(argc, argv)
	int argc;
	char *argv[];
{
	--argc; ++argv;
	while (argc > 0) {
		if (strcmp(argv[0], "-v") == 0) {
			++vflag;
			++argv, --argc;
			continue;
		}
		if (strcmp(argv[0], "-i") == 0 && argv[1] &&
		    isdigit(argv[1][0])) {
			interval = atoi(argv[1]);
			if (interval <= 0)
				usage();
			++argv, --argc;
			++argv, --argc;
			continue;
		}
		if (isdigit(argv[0][0])) {
			unit = atoi(argv[0]);
			if (unit < 0)
				usage();
			++argv, --argc;
			continue;
		}
		if (kflag)
			usage();

		system = *argv;
		++argv, --argc;
		if (argc > 0) {
			kmemf = *argv++;
			--argc;
			kflag++;
		}
	}
	if (kopen(system, kmemf, "slstats") < 0)
		exit(1);
	if (knlist(system, nl, "slstats") < 0)
		exit(1);
	intpr();
	exit(0);
}

#define V(offset) ((line % 20)? sc->offset - osc->offset : sc->offset)
#define AMT (sizeof(*sc) - 2 * sizeof(sc->sc_comp.tstate))

usage()
{
	static char umsg[] =
		"usage: slstats [-i interval] [-v] [unit] [system] [core]\n";

	fprintf(stderr, umsg);
	exit(1);
}

u_char	signalled;			/* set if alarm goes off "early" */

/*
 * Print a running summary of interface statistics.
 * Repeat display every interval seconds, showing statistics
 * collected over that interval.  Assumes that interval is non-zero.
 * First line printed at top of screen is always cumulative.
 */
intpr()
{
	register int line = 0;
	int oldmask;
	void catchalarm();
	struct sl_softc *sc, *osc;
	off_t addr;

	addr = nl[N_SOFTC].n_value + unit * sizeof(struct sl_softc);
	sc = (struct sl_softc *)malloc(AMT);
	osc = (struct sl_softc *)malloc(AMT);
	bzero((char *)osc, AMT);

	while (1) {
		if (kread(addr, (char *)sc, AMT) < 0)
			perror("kmem read");
		(void)signal(SIGALRM, catchalarm);
		signalled = 0;
		(void)alarm(interval);

		if ((line % 20) == 0) {
			printf("%8.8s %6.6s %6.6s %6.6s %6.6s",
				"IN", "PACK", "COMP", "UNCOMP", "ERR");
			if (vflag)
				printf(" %6.6s %6.6s", "TOSS", "IP");
			printf(" | %8.8s %6.6s %6.6s %6.6s %6.6s",
				"OUT", "PACK", "COMP", "UNCOMP", "IP");
			if (vflag)
				printf(" %6.6s %6.6s", "SEARCH", "MISS");
			putchar('\n');
		}
		printf("%8u %6d %6u %6u %6u",
			V(sc_if.if_ibytes),
			V(sc_if.if_ipackets),
			V(sc_comp.sls_compressedin),
			V(sc_comp.sls_uncompressedin),
			V(sc_comp.sls_errorin));
		if (vflag)
			printf(" %6u %6u",
				V(sc_comp.sls_tossed),
				V(sc_if.if_ipackets) -
				  V(sc_comp.sls_compressedin) -
				  V(sc_comp.sls_uncompressedin) -
				  V(sc_comp.sls_errorin));
		printf(" | %8u %6d %6u %6u %6u",
			V(sc_if.if_obytes),
			V(sc_if.if_opackets),
			V(sc_comp.sls_compressed),
			V(sc_comp.sls_packets) - V(sc_comp.sls_compressed),
			V(sc_if.if_opackets) - V(sc_comp.sls_packets));
		if (vflag)
			printf(" %6u %6u",
				V(sc_comp.sls_searches),
				V(sc_comp.sls_misses));

		putchar('\n');
		fflush(stdout);
		line++;
		oldmask = sigblock(sigmask(SIGALRM));
		if (! signalled) {
			sigpause(0);
		}
		sigsetmask(oldmask);
		signalled = 0;
		(void)alarm(interval);
		bcopy((char *)sc, (char *)osc, AMT);
	}
}

/*
 * Called if an interval expires before sidewaysintpr has completed a loop.
 * Sets a flag to not wait for the alarm.
 */
void
catchalarm()
{
	signalled = 1;
}

#include <kvm.h>
#include <fcntl.h>

kvm_t *kd;

kopen(system, kmemf, errstr)
	char *system;
	char *kmemf;
	char *errstr;
{
	if (strcmp(system, _PATH_UNIX) == 0 &&
	    strcmp(kmemf, _PATH_KMEM) == 0) {
		system = 0;
		kmemf = 0;
	}
	kd = kvm_open(system, kmemf, NULL, O_RDONLY, "slstats");
	if (kd == NULL) 
		return -1;
	return 0;
}

int
knlist(system, nl, errstr)
	char *system;
	struct nlist *nl;
	char *errstr;
{
	if (kd == 0)
		/* kopen() must be called first */
		abort();
		
	if (kvm_nlist(kd, nl) < 0 || nl[0].n_type == 0) {
		fprintf(stderr, "%s: %s: no namelist\n", errstr, system);
		return -1;
	}
	return 0;
}

int
kread(addr, buf, size)
	off_t addr;
	char *buf;
	int size;
{
	if (kvm_read(kd, (u_long)addr, buf, size) != size)
		return -1;
	return 0;
}
